File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        int maxArea = 0;
        Stack<int[]> stack = new Stack<>();
        for(int i = 0; i < n; i++){
            int start = i;
            while(!stack.isEmpty() && stack.peek()[1] > heights[i]){
                int[] ele = stack.pop();
                int idx = ele[0], height = ele[1];
                maxArea = Math.max(maxArea, height * (i - idx));
                start = idx;
            }
            stack.push(new int[]{start, heights[i]});
        }
        for(int[] ele: stack){
            int idx = ele[0];
            int height = ele[1];
            maxArea = Math.max(maxArea, height * (n - idx));
        }
        return maxArea;
    }
}

File: Data Structures & Algorithms/max-water-container/submission-2.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        maxWater = 0
        left, right = 0, len(heights) - 1
        while left < right:
            curr = (right - left) * min(heights[left], heights[right])
            maxWater = max(maxWater, curr)
            if heights[left] < heights[right]:
                left += 1
            else:
                right -= 1
        return maxWater
        